cpu/cc2538/mtarch.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2014, Institute for Pervasive Computing, ETH Zurich.
+ * All rights reserved.
+ *
+ * Author: Andreas Dröscher <contiki@anticat.ch>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "mtarch.h"
+#include "reg.h"
+#include "scb.h"
+#include "mt.h"
+
+#include <string.h>
+
+/**
+ * Pointer to the next thread or NULL if the next thread is main
+ */
+static struct mtarch_thread *next_thread = NULL;
+
+/**
+ * Pointer to the current thread or NULL the main thread is running
+ */
+static struct mtarch_thread *current_thread = NULL;
+
+/**
+ * Main stack pointer (used for kernel and protothreads)
+ */
+static void *msp = NULL;
+
+void
+mtarch_init(void)
+{
+}
+
+void
+mtarch_start(struct mtarch_thread *thread,
+             void (*function)(void *data),
+             void *data)
+{
+  /* Calculate SP */
+  mtarch_stack_t *sp = (mtarch_stack_t *)&thread->stack[MTARCH_STACKSIZE - MTARCH_REGISTER_COUNT];
+  thread->sp = sp;
+
+  /* Zero stack */
+  memset(thread->stack, '\0', sizeof(void *) * MTARCH_STACKSIZE);
+
+  /* Fill stack */
+  sp->r0 = data;
+  sp->lr = mt_exit;
+  sp->pc = function;
+  sp->psr = (void *)0x01000000;
+}
+
+void
+mtarch_yield(void)
+{
+  /* Trigger PendSV the handler will switch to main */
+  REG(SCB_INTCTRL) = SCB_INTCTRL_PEND_SV;
```

Comment:
You should rather use `svc` here. The `SVCall` handler would always switch to the main thread (kernel + protothreads) using `msp`, and the `PendSV` handler would always switch to the process thread using `psp`. This would also make it possible to use the `mt` stuff in unprivileged mode.

In that way, you could remove most of your state variables, which would also remove the duplication of the current thread pointer with `mt.c`.

Using the banked versions of `sp` also gets rid of your `msp` variable.